src/main/scala/format/pgn/Node.scala

```diff
@@ -0,0 +1,101 @@
+package chess
+package format
+package pgn
+
+import cats.syntax.all.*
+import cats.*
+
+import monocle.{ Optional, Traversal }
+
+/**
+ * Node is a tree structure specialized for chess games.
+ * It's a tree with a main line and variations.
+ * Node is a [Functor](https://typelevel.org/cats/typeclasses/functor.html)
+ * Node is a [Traverse](https://typelevel.org/cats/typeclasses/traverse.html)
+ * It is also provided some Monocle optics see Node.filterOptional and Node.filterTraversal
+  */
+case class Node[A](
+    value: A,
+    child: Option[Node[A]],   // main line next move
+    variations: List[Node[A]] // alternate moves, e.g. Node("e4", None, List("d4", "Nf3"))
+):
+  def mainLine: List[A] = value :: child.fold(Nil)(_.mainLine)
+  def totalNodes: Int   = this.foldLeft(0)((b, _) => b + 1)
+
+  // find the first Node that statisfies the predicate
+  def findNode(predicate: A => Boolean): Option[Node[A]] =
+    if predicate(value) then this.some
+    else
+      child
+        .flatMap(_.findNode(predicate))
+        .orElse(variations.foldLeft(none[Node[A]])((b, n) => b.orElse(n.findNode(predicate))))
+
+  def replaceNode(predicate: A => Boolean)(node: Node[A]): Option[Node[A]] =
+    modifyNode(predicate)(_ => node)
+
+  // modify the first node that satisfies the predicate (dfs with main line and then variations)
+  def modifyNode(predicate: A => Boolean)(f: Node[A] => Node[A]): Option[Node[A]] =
+    if predicate(value) then f(this).some
+    else
+      child.flatMap(_.modifyNode(predicate)(f)) match
+        case Some(n) => copy(child = Some(n)).some
+        case None =>
+          variations.foldLeft((false, List.empty[Node[A]])) {
+            case ((true, acc), n) => (true, acc :+ n)
+            case ((false, acc), n) =>
+              n.modifyNode(predicate)(f) match
+                case Some(nn) => (true, acc :+ nn)
+                case None     => (false, acc :+ n)
+          } match
+            case (true, ns) => copy(variations = ns).some
+            case (false, _) => none
+
+  // delete the first node that satisfies the predicate (both child and variations)
+  // except the root
+  def deleteSubNode(predicate: A => Boolean): Option[Node[A]] =
+    child.flatMap { n =>
+      if predicate(n.value) then copy(child = None).some
+      else n.deleteSubNode(predicate).map(nn => this.copy(child = Some(nn)))
+    } match
+      case Some(n) => n.some
+      case None =>
+        variations.foldLeft((false, List.empty[Node[A]])) {
+          case ((true, acc), n) => (true, acc :+ n)
+          case ((false, acc), n) =>
+            if predicate(n.value) then (true, acc)
+            else
+              n.deleteSubNode(predicate) match
+                case Some(nn) => (true, acc :+ nn)
+                case None     => (false, acc :+ n)
+        } match
+          case (true, ns) => copy(variations = ns).some
+          case (false, _) => none
```

Comment:
Same question about variation storage in the `acc`. Also I wonder if we can factor out `modifyNode` and `deleteSubnote` by having a modified `f: Node => Option[Node]`. I've tried to write it here and couldn't make it without compiler at hand though.

Reply:
interesting, If I understand correctly, you suggest to have a function:
`superPower(predicate: A => Boolean)(f: Node => Option[Node])`
if f returns None means delete the node, and then we can rewrite, modify and delete with new `superPower` function?

I can try to implement that.

Reply:
That was the idea yeah.